runtime/eventring_consumer.c

```diff
@@ -0,0 +1,686 @@
+             msg_length * sizeof(uint64_t));
+
+      ring_head = atomic_load_explicit(&eventring_buffer_header->ring_head,
+                                       memory_order_acquire);
+
+      /* Check the message we've read hasn't been overwritten by the writer */
+      if (ring_head > cursor->current_positions[domain_num]) {
+        /* It potentially has, retry for the next one after we've notified
+             the callbacks about lost messages. */
+        int lost_words = ring_head - cursor->current_positions[domain_num];
+        cursor->current_positions[domain_num] = ring_head;
+
+        if (cursor->lost_events) {
+          if( !(cursor->lost_events(domain_num, callback_data, lost_words)) ) {
+            early_exit = 1;
+            continue;
+          }
+
+        }
+      }
+
+      switch (EVENTRING_ITEM_TYPE(header)) {
+      case EV_BEGIN:
+        if (cursor->runtime_begin) {
+          if( !cursor->runtime_begin(domain_num, callback_data, buf[1],
+                                      EVENTRING_ITEM_ID(header)) ) {
+                                        early_exit = 1;
+                                        continue;
+                                      }
+        }
+        break;
+      case EV_EXIT:
+        if (cursor->runtime_end) {
+          if( !cursor->runtime_end(domain_num, callback_data, buf[1],
+                                    EVENTRING_ITEM_ID(header)) ) {
+                                      early_exit = 1;
+                                      continue;
+                                    };
+        }
+        break;
+      case EV_COUNTER:
+        if (cursor->runtime_counter) {
+          if( !cursor->runtime_counter(domain_num, callback_data, buf[1],
+                                        EVENTRING_ITEM_ID(header), buf[2]) ) {
+                                          early_exit = 1;
+                                          continue;
+                                        };
+        }
+        break;
+      case EV_ALLOC:
+        if (cursor->alloc) {
+          if( !cursor->alloc(domain_num, callback_data, buf[1], &buf[2])) {
+            early_exit = 1;
+            continue;
+          }
+        }
+        break;
+      case EV_LIFECYCLE:
+        if (cursor->lifecycle) {
+          if( !cursor->lifecycle(domain_num, callback_data, buf[1],
+                                  EVENTRING_ITEM_ID(header), buf[2]) ) {
+                                    early_exit = 1;
+                                    continue;
+                                  }
+        }
+      }
+
+      if (EVENTRING_ITEM_TYPE(header) != EV_INTERNAL) {
+        consumed++;
+      }
+
+      cursor->current_positions[domain_num] += msg_length;
+    } while (cursor->current_positions[domain_num] < ring_tail &&
+             (max_events == 0 || consumed < max_events) && !early_exit);
+
+    cursor->next_read_domain = (domain_num + 1) % cursor->metadata->max_domains;
```

Comment:
Modulus is expensive. Assuming you can't guarantee `cursor->metadata->max_domains` is divisible by two then might be worth replacing with `domain_num + 1 == cursor->metadata->max_domains ? 0 : domain_num + 1`?